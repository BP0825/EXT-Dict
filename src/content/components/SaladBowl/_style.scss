@import '@/_sass_global/variables';
@import '@/_sass_global/z-indices';
@import '@/_sass_global/interfaces';

$bowl-width: 30px;
$tomato-rotate: 45deg;
$leaf-rotate: 30deg;

:root:root:root:root:root {
  .saladict-SaladBowl {
    @extend %reset-important;
    position: fixed !important;
    z-index: $global-zindex-tooltip !important;
    left: 0 !important;
    top: 0 !important;
    overflow: hidden !important;
    width: $bowl-width !important;
    height: $bowl-width !important;
    user-select: none !important;
  }

  .saladict-SaladBowl_Leaf {
    @extend %reset-important;
    position: absolute !important;
    z-index: 1 !important;
    top: 50 * $bowl-width / 1024 !important;
    left: 600 * $bowl-width / 1024 !important;
    width: $bowl-width * 729 / 1024 * 0.6 !important;
    height: $bowl-width * 0.6 !important;
    transform: rotate($leaf-rotate) !important;
  }

  .saladict-SaladBowl_Orange {
    @extend %reset-important;
    position: absolute !important;
    z-index: 2 !important;
    top: 0 * $bowl-width / 1024 !important;
    left: 195 * $bowl-width / 1024 !important;
    width: $bowl-width * 0.6 !important;
    height: $bowl-width * 0.6 !important;
  }

  .saladict-SaladBowl_Tomato {
    @extend %reset-important;
    position: absolute !important;
    z-index: 3 !important;
    top: 250 * $bowl-width / 1024 !important;
    left: -117 * $bowl-width / 1024 !important;
    width: $bowl-width * 0.6 !important;
    height: $bowl-width * 513 / 1024 * 0.6 !important;
    transform: rotate($tomato-rotate) !important;
  }

  .saladict-SaladBowl_Bowl {
    @extend %reset-important;
    position: absolute !important;
    z-index: 4 !important;
    bottom: 0 !important;
    left: 0 !important;
    width: $bowl-width !important;
    height: $bowl-width * 530 / 1024 !important;
  }
}

/*-----------------------------------------------*\
    States
\*-----------------------------------------------*/

.saladict-SaladBowl:hover {
  .saladict-SaladBowl_Leaf {
    animation: saladict-SaladBowl_Leaf-shake 1s infinite linear;
  }
  .saladict-SaladBowl_Orange {
    animation: saladict-SaladBowl_Orange-spin 1s infinite linear;
  }
  .saladict-SaladBowl_Tomato {
    animation: saladict-SaladBowl_Tomato-shake 0.7s infinite linear;
  }
}

@keyframes saladict-SaladBowl_Leaf-shake {
  0% { transform: translate( 2px,  1px) rotate($leaf-rotate + 0deg); }
 10% { transform: translate(-1px, -2px) rotate($leaf-rotate - 1deg); }
 20% { transform: translate(-2px,  0  ) rotate($leaf-rotate + 1deg); }
 30% { transform: translate( 0  ,  2px) rotate($leaf-rotate + 0deg); }
 40% { transform: translate( 1px, -1px) rotate($leaf-rotate + 1deg); }
 50% { transform: translate(-1px,  2px) rotate($leaf-rotate - 1deg); }
 60% { transform: translate(-2px,  1px) rotate($leaf-rotate + 0deg); }
 70% { transform: translate( 2px,  1px) rotate($leaf-rotate - 1deg); }
 80% { transform: translate(-1px, -1px) rotate($leaf-rotate + 1deg); }
 90% { transform: translate( 2px,  2px) rotate($leaf-rotate + 0deg); }
100% { transform: translate( 1px, -2px) rotate($leaf-rotate - 1deg); }
}

@keyframes saladict-SaladBowl_Orange-spin {
from { transform: rotate(  0deg); }
  to { transform: rotate(360deg); }
}

@keyframes saladict-SaladBowl_Tomato-shake {
  0% { transform: rotate($tomato-rotate - 10deg); }
 30% { transform: rotate($tomato-rotate +  0deg); }
 60% { transform: rotate($tomato-rotate + 10deg); }
 90% { transform: rotate($tomato-rotate +  0deg); }
100% { transform: rotate($tomato-rotate -  5deg); }
}
