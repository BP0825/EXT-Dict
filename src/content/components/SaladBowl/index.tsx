import './_style.scss'
import React from 'react'
import { Spring, SpringConfig } from 'react-spring'

const imgLeaf = require('@/assets/leaf.svg')
const imgOrange = require('@/assets/orange.svg')
const imgTomato = require('@/assets/tomato.svg')
const imgBowl = require('@/assets/bowl.svg')

export type SaladBowlProps = {
  readonly x: number
  readonly y: number
  readonly scale: number
  readonly springImmediateCtrl: (key: string) => boolean
  readonly springConfigCtrl: (key: string) => SpringConfig
  readonly mouseOnBowl: (flag: boolean) => any
  readonly searchText: () => any
}

export default class SaladBowl extends React.PureComponent<SaladBowlProps> {
  readonly bowlRef = React.createRef<HTMLDivElement>()
  initStyle = { x: 0, y: 0, scale: 0 }
  mouseOnBowlTimeout: any

  handleMouseEnter = () => {
    this.mouseOnBowlTimeout = setTimeout(() => {
      this.props.mouseOnBowl(true)
      this.props.searchText()
    }, 800)
  }

  handleMouseLeave = () => {
    clearTimeout(this.mouseOnBowlTimeout)
    this.props.mouseOnBowl(false)
  }

  animateBowl = style => {
    const el = this.bowlRef.current
    if (el) {
      const { x, y, scale } = style
      el.style.setProperty(
        'transform',
        `translate3d(${x}px, ${y}px, 0) scale(${scale})`,
        'important',
      )
    }
    return null
  }

  render () {
    const { x, y, scale, springConfigCtrl, springImmediateCtrl } = this.props
    return (
      <div className='saladict-SaladBowl'
        key={'saladict-SaladBowl'}
        ref={this.bowlRef}
        onMouseEnter={this.handleMouseEnter}
        onMouseLeave={this.handleMouseLeave}
      >
        <Spring
          from={this.initStyle}
          to={{ x, y, scale }}
          immediate={springImmediateCtrl}
          config={springConfigCtrl}
        >{this.animateBowl}</Spring>
        <img className='saladict-SaladBowl_Leaf' src={imgLeaf} />
        <img className='saladict-SaladBowl_Orange' src={imgOrange} />
        <img className='saladict-SaladBowl_Tomato' src={imgTomato} />
        <img className='saladict-SaladBowl_Bowl' src={imgBowl} />
      </div>
    )
  }
}
