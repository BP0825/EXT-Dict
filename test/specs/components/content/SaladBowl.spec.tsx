import React from 'react'
import renderer from 'react-test-renderer'

import SaladBowl from '@/content/components/SaladBowl'

describe('Component/content/SaladBowl', () => {
  it('should render correctly', () => {
    const props = {
      x: 0,
      y: 0,
      scale: 1,
    }
    const tree = renderer.create(<SaladBowl {...props}/>).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
