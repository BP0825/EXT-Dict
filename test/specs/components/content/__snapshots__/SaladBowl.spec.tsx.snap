// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component/content/SaladBowl should render correctly 1`] = `
<div
  className="saladict-SaladBowl isAnimate"
  key="saladict-SaladBowl"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <img
    className="saladict-SaladBowl_Leaf"
    src="leaf.svg"
  />
  <img
    className="saladict-SaladBowl_Orange"
    src="orange.svg"
  />
  <img
    className="saladict-SaladBowl_Tomato"
    src="tomato.svg"
  />
  <img
    className="saladict-SaladBowl_Bowl"
    src="bowl.svg"
  />
</div>
`;
